mapie/multi_label_classification.py

```diff
@@ -0,0 +1,795 @@
+            First variance in the sigma_hat array. The default
+            value is the same as in the paper implementation.
+
+            By default .25
+
+        Returns
+        -------
+        Tuple[NDArray, NDArray] of shape (n_lambdas, ) and (n_lambdas)
+            Average risk over all the obervations and upper bound of the risk.
+        """
+        r_hat = self.risks.mean(axis=0)
+        n_obs = len(self.risks)
+
+        if (self.method == "rcps") and (delta is not None):
+            if bound == "hoeffding":
+                r_hat_plus = (
+                    r_hat +
+                    np.sqrt((1 / (2 * n_obs)) * np.log(1 / delta))
+                )
+
+            elif bound == "bernstein":
+                sigma_hat_bern = np.var(r_hat, axis=0, ddof=1)
+                r_hat_plus = (
+                    r_hat +
+                    np.sqrt((sigma_hat_bern * 2 * np.log(2 / delta)) / n_obs) +
+                    (7 * np.log(2 / delta)) / (3 * (n_obs - 1))
+                )
+
+            else:
+                mu_hat = (
+                    (.5 + np.cumsum(self.risks, axis=0)) /
+                    (np.repeat(
+                        [range(1, n_obs + 1)],
+                        self.n_lambdas,
+                        axis=0
+                    ).T + 1)
+                )
+                sigma_hat = (
+                    (.25 + np.cumsum((self.risks - mu_hat)**2, axis=0)) /
+                    (np.repeat(
+                        [range(1, n_obs + 1)],
+                        self.n_lambdas,
+                        axis=0
+                    ).T + 1)
+                )
+                sigma_hat = np.concatenate(
+                    [
+                        np.full(
+                            (1, self.n_lambdas), fill_value=self.sigma_init
+                        ) * self.sigma_init, sigma_hat[:-1]
+                    ]
+                )
+                nu = np.minimum(
+                    1,
+                    np.sqrt((2 * np.log(1 / delta)) / (n_obs * sigma_hat))
+                )
+
+                # Split the calculation in two to prevent memory issues
+                batches = [
+                    range(int(n_obs / 2)),
+                    range(n_obs - int(n_obs / 2), n_obs)
+                ]
+                K_R_max = np.zeros((self.n_lambdas, self.n_lambdas))
+                for batch in batches:
+                    nu_batch = nu[batch]
+                    losses_batch = self.risks[batch]
+
+                    nu_batch = np.repeat(
+                        np.expand_dims(nu_batch, axis=2),
+                        self.n_lambdas,
+                        axis=2
+                    )
+                    losses_batch = np.repeat(
+                        np.expand_dims(losses_batch, axis=2),
+                        self.n_lambdas,
+                        axis=2
+                    )
+
+                    R = np.arange(self.n_lambdas) / self.n_lambdas
```

Comment:
use the store self.lambdas : much more compact and easy to understand

Reply:
done